### ISSUE TITLE
Step 4: Detailed Operation

### ISSUE THREAD

**Issue by sunminghao4536** [yininghase/multiagent-collision-mining#2]
![image](https://github.com/user-attachments/assets/6e7c51b9-0302-4815-9f98-dbb0154a94a3)I ran calculate_metrics.py in the third step to generate metrics.pt, and then I didn't understand the fourth step, because mpc.py here is called in simulation.py, and then simulation.py is called in generate_trainval_data.py and reference.py. May I ask how to implement the fourth step of obtaining extended training labels。Thanks

**Comment by yininghase** [comment]
Yes, the mpc.py should be called by generate_trainval_data.py. 

For step 4, the metrics.pt records the collision state for each trajectory. Pick up those trajectories with a high collision rate. The data structure for the picked-up trajectories should also contain batches_data, X_data, y_model_data and trajectory_data. 

For each data point in these trajectories, you should use MPC to get the ground truth label by running generate_trainval_data.py. Modify the 'control init' to the picked-up data folder in the config file so that this picked-up data will be loaded in generate_trainval_data.py. 

Note that, you do not need to drive the vehicles to their targets in this step. By setting the simulation time to 1, the MPC will only run for the current data point without moving the vehicles for the next time step.

**Comment by sunminghao4536** [comment]
Thank you very much, happy

**Comment by sunminghao4536** [comment]
Hello, in the third step, I ran calculate_tetrics.py to generate metrics.pt, and then directly ran generate_trainval_data-py to generate training labels for only 8 vehicles and 0 obstacles. The running results are shown in the following figure
![image](https://github.com/user-attachments/assets/bb8facce-7ee4-48ca-8a4d-4a9173dc0f06)
May I ask if the timing of this training data is normal? Am I missing a step in the third step (selecting trajectories with high collision rates as difficult scenarios) when I ran calculate_tetrics.py to generate metrics.pt? What should be done if this step is missing?

**Comment by yininghase** [comment]
Yes, because you generate ground truth data on all trajectories and it would take much more time. As the current code will not automatically select those trajectories with a high collision rate, you may need to do it manually on your own. 

To select the hard trajectories, you may need to know how the data are organized in batches_data, X_data, y_model_data and trajectory_data.pt. Relevant code is shown in the comment of simulation.py from line 162 to 192. Here, I can also briefly explain this with an example. Suppose there are 3 trajectories with m vehicles and n obstacles. The time length for these 3 trajectories are t1, t2 and t3: 

1. The X_data.pt saves the vehicle/obstacle states of these trajectories with a tensor shape of ((t1+t2+t3)*(m+n), 8). In X_data.pt, every (m+n) rows save the states of the m vehicles and n obstacles for a specific moment. As the first trajectory has a timestep of t1, the first t1*(m+n) lines save the movements of the m vehicles and n obstacles of the first trajectories. The rows from t1*(m+n)+1 to (t1+t2)*(m+n) are for the second trajectories. The rows from (t1+t2)*(m+n)+1 to (t1+t2+t3)*(m+n) are for the third. Within each row: 
        # For each data point of vehicle we have a vehicle position (x,y), angle and velocity and the desired position and angle, 
        # For each data point of obstacle we have a obstacle position (x,y), radius 
        # To distinguish the vehicle and obstacle, we add a sign addtionally, for vehicle it is 0, for obstacle it is 1
        # so for vehicle: [x, y, angle, v, x_d, y_d, angle_d, 0], for obstacle: [0, 0, 0, 0, x, y, r, 1]
        # for the problem of m vehicles and n obstacles, we stack the first m vehicles and n obstacles together like
        # [[x, y, angle, v, x_d, y_d, angle_d, 0], # vehicle 1 
        #             ......
        #  [x, y, angle, v, x_d, y_d, angle_d, 0], # vehicle m
        #  [0, 0, 0, 0, x, y, r, 1],   # obstacle 1
        #             ......
        #  [0, 0, 0, 0, x, y, r, 1],   # obstacle n
        # ]

2.  The y_data saves vehicle controls with a tensor shape of ((t1+t2+t3)*m, 2*N). Because only the m vehicles need to be controlled and the n obstacles do not. In this case, the first t1*m lines save the controls of the m vehicles. The rows from t1*m+1 to (t1+t2)*m are for the second and the rows from (t1+t2)*m+1 to (t1+t2+t3)*m are for the third. Within each row:
        # Each ground truth label of vehicle data points has to contain the predicted steps of 
        # pedal and steering angle at each step of the horizon N, like
        # [[pedal_0, steering_angle_0, pedal_1, steering_angle_1, ..., pedal_N, steering_angle_N]  # vehicle 1 
        #  ...
        #  [pedal_0, steering_angle_0, pedal_1, steering_angle_1, ..., pedal_N, steering_angle_N]  # vehicle m
        # ]
Note that, for MPC there are prediction horizon of N, but the GNN only predict the next step so N=1.

3. The batch_data saves a tensor with shape of ((t1+t2+t3), 2) being:
        # [[m+n, m],
        #        ....      ,
        #  [m+n, m]]

4. The trajectory_data saves a tensor with a shape of (4,1), indicating the trajectory index of X_data.pt:
        # [0, t1*(m+n), (t1+t2)*(m+n), (t1+t2+t3)*(m+n)]

After you are familiar with the data structure you can operate these tensor files to pick up your desired data points. For example, if I want to pick up the data point in trajectory 2 from step n1 to step n2:

1. Read the X_data.pt -> Reshape the tensor to ((t1+t2+t3), (m+n), 8) -> Clip the segment from t1+n1 to t2+n2 ->Reshape the tensor back to ((n2-n1)*(m+n), 8).

2. Read the y_data.pt -> Reshape the tensor to ((t1+t2+t3), (m+n), 2*N) -> Clip the segment from t1+n1 to t2+n2 ->Reshape the tensor back to ((n2-n1)*(m+n), 2*N)

3. The batch_data.pt is directly:   [[m+n, m], ..., [m+n, m]] with (n2-n1) rows.

4. The trajectory_data.pt is  [0, (n2-n1)*(m+n)]

Then, which desired data points are regarded as hard data? This information is given by the "collision times" and "travel distance" in metrics.pt. How they are calculated is shown in the calculate_matrics.py. Generally,  "collision times" records how many collisions happen in the trajectory. "travel distance" records the length of the trajectory. Calculate the collision rate for each trajectory (collision_time/travel_distance) and pick up those trajectories with high collision rate. Only these trajectories are used for ground truth generation. By the way, you may not use the whole trajectory but only those segments before the vehicle collision.
